velox/expression/ComplexProxyTypes.h

```diff
@@ -0,0 +1,222 @@
+  FOLLY_ALWAYS_INLINE element_t& add_item() {
+    commitMostRecentChildItem();
+    auto index = valuesOffset_ + length_;
+    reserve(index + 1);
+    length_++;
+
+    if constexpr (!provide_std_interface) {
+      childWriter_->setOffset(index);
+      needCommit_ = true;
+      return childWriter_->current();
+    } else {
+      childWriter_->vector().setNull(index, false);
+      return childWriter_->data_[index];
+    }
+  }
+
+  // Add a new null item to the array.
+  FOLLY_ALWAYS_INLINE void add_null() {
+    commitMostRecentChildItem();
+    auto index = valuesOffset_ + length_;
+    reserve(index + 1);
+    length_++;
+    childWriter_->vector().setNull(index, true);
+    // Note: no need to commit the null item.
+  }
+
+  // Should be called by the user (VectorWriter) when writing is done to commit
+  // last item if needed.
+  void finalize() {
+    commitMostRecentChildItem();
+    // Downsize to the actual size used in the underlying vector.
+    // Some vector-writer's logic depend on the previous size to append data.
+    childWriter_->vector().resize(valuesOffset_ + length_);
+  }
+
+  vector_size_t size() {
+    return length_;
+  }
+
+  // Functions below provide an std::like interface, and are enabled only when
+  // the array element is primitive that is not string or bool.
+
+  // 'size' is with respect to the current size of the array being written.
+  FOLLY_ALWAYS_INLINE typename std::enable_if<provide_std_interface>::type
+  resize(vector_size_t size) {
+    commitMostRecentChildItem();
+    reserve(size);
+    length_ = size;
+  }
+
+  typename std::enable_if<provide_std_interface>::type FOLLY_ALWAYS_INLINE
+  push_back(element_t value) {
+    auto& item = add_item();
+    item = value;
+  }
+
+  typename std::enable_if<provide_std_interface>::type FOLLY_ALWAYS_INLINE
+  push_back(std::nullopt_t) {
+    add_null();
+  }
+
+  typename std::enable_if<provide_std_interface>::type FOLLY_ALWAYS_INLINE
+  push_back(const std::optional<element_t>& value) {
+    if (value) {
+      push_back(*value);
+    } else {
+      add_null();
+    }
+  }
+
+  typename std::enable_if<provide_std_interface, PrimitiveWriterProxy<V>>::type
+  operator[](vector_size_t index_) {
+    return PrimitiveWriterProxy<V>{
+        &childWriter_->vector(), valuesOffset_ + index_};
+  }
+
+ private:
+  ArrayProxy<V>() {}
+
+  FOLLY_ALWAYS_INLINE void commitMostRecentChildItem() {
+    if constexpr (!provide_std_interface) {
+      if (needCommit_) {
+        childWriter_->commit(true);
+        needCommit_ = false;
+      }
+    }
+  }
+
+  // Prepare the proxy for a new element.
+  FOLLY_ALWAYS_INLINE void init(vector_size_t valuesOffset) {
+    valuesOffset_ = valuesOffset;
+    length_ = 0;
+    capacity_ = 0;
+    needCommit_ = false;
+  }
+
+  void setChildWriter(child_writer_t* childWriter) {
+    childWriter_ = childWriter;
+  }
+
+  // Pointer to child vector writer.
+  child_writer_t* childWriter_;
```

Comment:
These variables are all uninitialized -- I would strongly prefer they were initialized to stable values, otherwise behavior may vary by optimization level.